packages/hoppscotch-js-sandbox/src/utils.ts

```diff
@@ -1,89 +1,439 @@
+        resolve: envResolveFn,
+      },
+    },
+    updatedEnvs,
+  }
 }
 
-export function setEnv(
+const setEnv = (
   envName: string,
   envValue: string,
   envs: TestResult["envs"]
-): TestResult["envs"] {
-  const indexInSelected = envs.selected.findIndex((x) => x.key === envName)
+): TestResult["envs"] => {
+  const { global, selected } = envs
+
+  const indexInSelected = selected.findIndex(
+    (x: SelectedEnvItem) => x.key === envName
+  )
 
   // Found the match in selected
   if (indexInSelected >= 0) {
-    envs.selected[indexInSelected].value = envValue
+    selected[indexInSelected].value = envValue
 
     return {
-      global: envs.global,
-      selected: envs.selected,
+      global,
+      selected,
     }
   }
 
-  const indexInGlobal = envs.global.findIndex((x) => x.key == envName)
+  const indexInGlobal = global.findIndex((x: GlobalEnvItem) => x.key == envName)
 
   // Found a match in globals
   if (indexInGlobal >= 0) {
-    envs.global[indexInGlobal].value = envValue
+    global[indexInGlobal].value = envValue
 
     return {
-      global: envs.global,
-      selected: envs.selected,
+      global,
+      selected,
     }
   }
 
   // Didn't find in both places, create a new variable in selected
-  envs.selected.push({
+  selected.push({
     key: envName,
     value: envValue,
   })
 
   return {
-    global: envs.global,
-    selected: envs.selected,
+    global,
+    selected,
+  }
+}
+
+export function preventCyclicObjects(
+  obj: Record<string, any>
+): E.Left<string> | E.Right<Record<string, any>> {
+  let jsonString
+
+  try {
+    jsonString = JSON.stringify(obj)
+  } catch (e) {
+    return E.left("Stringification failed")
+  }
+
+  try {
+    const parsedJson = JSON.parse(jsonString)
+    return E.right(parsedJson)
+  } catch (err) {
+    return E.left("Parsing failed")
+  }
+}
+
+/**
+ * Creates an Expectation object for use inside the sandbox
+ * @param expectVal The expecting value of the expectation
+ * @param negated Whether the expectation is negated (negative)
+ * @param currTestStack The current state of the test execution stack
+ * @returns Object with the expectation methods
+ */
+export const createExpectation = (
+  expectVal: any,
+  negated: boolean,
+  currTestStack: TestDescriptor[]
+) => {
+  const result: Record<string, unknown> = {}
+
+  const toBeFn = (expectedVal: any) => {
+    let assertion = expectVal === expectedVal
+
+    if (negated) {
+      assertion = !assertion
+    }
+
+    const status = assertion ? "pass" : "fail"
+    const message = `Expected '${expectVal}' to${
+      negated ? " not" : ""
+    } be '${expectedVal}'`
+
+    currTestStack[currTestStack.length - 1].expectResults.push({
+      status,
+      message,
+    })
+
+    return undefined
+  }
+
+  const toBeLevelXxx = (
+    level: string,
+    rangeStart: number,
+    rangeEnd: number
+  ) => {
+    const parsedExpectVal = parseInt(expectVal)
+
+    if (!Number.isNaN(parsedExpectVal)) {
```

Comment:
```js
pw.expect("404").toBeLevel4xx();
```

Added support for the expect value being a `string` with `toBeLevel` matchers. It was mentioned in the [documentation](https://docs.hoppscotch.io/documentation/features/scripts#pwtobelevelxxx) such that it was supported already.